### PR TITLE
fix post launch call

### PR DESCRIPTION
**Pull request by artv3**
Addresses https://github.com/LLNL/RAJA/pull/1393#discussion_r1037285613

### CODE REVIEW

**Review by rhornung67 — APPROVED** [review]

**Comment by MrBurmark** [comment]
The ABORT_OR_THROW will now always be run too.

**Comment by rhornung67** [comment]
> The ABORT_OR_THROW will now always be run too.

Folks, I have several short fuse items I need to tackle today. Please work this out and merge when it's ready.

File: include/RAJA/pattern/launch/launch_core.hpp [hunk 1]
```diff
@@ -355,21 +355,23 @@ launch(RAJA::resources::Resource res, LaunchParams const &params, const char *ke
...
+      util::callPostLaunchPlugins(context);
+      return e_proxy;
     }
 #ifdef RAJA_DEVICE_ACTIVE
     case ExecPlace::DEVICE: {
       using launch_t = LaunchExecute<typename POLICY_LIST::device_policy_t>;
-      return launch_t::exec(res, params, kernel_name, p_body); break;
+      resources::EventProxy<resources::Resource> e_proxy = launch_t::exec(res, params, kernel_name, p_body); break;
+      util::callPostLaunchPlugins(context);
+      return e_proxy;
```
Comment (artv3) [inline include/RAJA/pattern/launch/launch_core.hpp:367]: Is there any overhead for making these copies?

Comment (MrBurmark) [reply include/RAJA/pattern/launch/launch_core.hpp:367]: I think this should optimize out.

**Review by MrBurmark — APPROVED** [review]

File: include/RAJA/pattern/launch/launch_core.hpp [hunk 2]
```diff
@@ -355,21 +355,23 @@ launch(RAJA::resources::Resource res, LaunchParams const &params, const char *ke
...
     }
 #ifdef RAJA_DEVICE_ACTIVE
     case ExecPlace::DEVICE: {
       using launch_t = LaunchExecute<typename POLICY_LIST::device_policy_t>;
-      return launch_t::exec(res, params, kernel_name, p_body); break;
+      resources::EventProxy<resources::Resource> e_proxy = launch_t::exec(res, params, kernel_name, p_body); break;
```
Comment (MrBurmark) [inline include/RAJA/pattern/launch/launch_core.hpp:365]: remove the breaks on these lines

Comment (artv3) [inline include/RAJA/pattern/launch/launch_core.hpp:365]:
```suggestion
      resources::EventProxy<resources::Resource> e_proxy = launch_t::exec(res, params, kernel_name, p_body);
```

File: include/RAJA/pattern/launch/launch_core.hpp [hunk 3]
```diff
@@ -355,21 +355,23 @@ launch(RAJA::resources::Resource res, LaunchParams const &params, const char *ke
   switch (place) {
     case ExecPlace::HOST: {
       using launch_t = LaunchExecute<typename POLICY_LIST::host_policy_t>;
-      return launch_t::exec(res, params, kernel_name, p_body); break;
+      resources::EventProxy<resources::Resource> e_proxy = launch_t::exec(res, params, kernel_name, p_body); break;
```
Comment (artv3) [inline include/RAJA/pattern/launch/launch_core.hpp:358]:
```suggestion
      resources::EventProxy<resources::Resource> e_proxy = launch_t::exec(res, params, kernel_name, p_body);
```

**Comment by MrBurmark** [comment]
Does the launch plugin test test this code path?

**Comment by artv3** [comment]
> Does the launch plugin test test this code path?

No, we only have coverage for the non-resource version. I can add some here though.

**Review by rhornung67 — APPROVED** [review]

**Comment by artv3** [comment]
I added test for the launch, resource, plugin code path. I also reduced duplication of the non-resource version launch API. Let me know if these changes are okay.

File: test/integration/plugin/tests/test-plugin-resource-launch.hpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  for (int i = 0; i < 10; i++) {
+
+    //Keep PluginTestCallable within a scope to ensure
+    //destruction, consistent with other test
+    {
+      PluginTestCallable p_callable{data};
+
+      RAJA::launch<LaunchPolicy>
+        (WORKING_RES::get_default(), RAJA::LaunchParams(RAJA::Teams(1), RAJA::Threads(1)),
```
Comment (artv3) [inline test/integration/plugin/tests/test-plugin-resource-launch.hpp:41]: Is this okay? or do we want to try the non-default option?

Comment (rhornung67) [reply test/integration/plugin/tests/test-plugin-resource-launch.hpp:41]: I think testing all cases that you think users may want is what we should do.

**Review by MrBurmark — APPROVED** [review]

File: test/integration/plugin/tests/test-plugin-resource-launch.hpp (hunk 4 above)
Comment (MrBurmark) [reply test/integration/plugin/tests/test-plugin-resource-launch.hpp]: I don't think which value of the resource matters as long as there is a test that exercises the resource code path.

**Pull request merged**